src/server_manager/web_app/app.ts

```diff
@@ -200,136 +209,218 @@ export class App {
+            this.maybeShowLastDisplayedServer();
+          });
+        });
       }
+    });
+    this.showIntro();
+  }
 
-      // User has no manual servers or DigitalOcean token.
-      this.showIntro();
+  private async syncServersToDisplay(servers: server.Server[]) {
+    for (const server of servers) {
+      await this.syncServerToDisplay(server);
+    }
+  }
+
+  // Syncs the locally persisted server metadata for `server`. Creates a DisplayServer for `server`
+  // if one is not found in storage.
+  private async syncServerToDisplay(server: server.Server): Promise<DisplayServer> {
+    const displayServerId = server.getManagementApiUrl();
+    let displayServer = this.displayServerRepository.findServer(displayServerId);
+    if (!displayServer) {
+      console.warn(`Could not find managed display server with ID ${displayServerId}`);
+      const isHealthy = await server.isHealthy().catch((e) => false);
+      displayServer = {
+        id: displayServerId,
+        name: isHealthy ? server.getName() : server.getHostname(),
+        isManaged: isManagedServer(server)
+      };
+      this.displayServerRepository.addServer(displayServer);
+      this.syncDisplayServersToUi();
+    } else {
+      // We may need to update the stored display server if it was persisted when the server was not
+      // healthy.
+      try {
+        const remoteServerName = server.getName();
+        if (displayServer.name !== remoteServerName) {
+          displayServer.name = remoteServerName;
+          this.removeServerFromDisplay(displayServer);
+          this.displayServerRepository.addServer(displayServer);
+          this.syncDisplayServersToUi();
+        }
+      } catch (e) {
+        // Ignore, we may not have the server config yet.
+      }
+    }
+    this.serverDisplayTable[displayServer.id] = {displayServer, server};
+    return displayServer;
+  }
+
+  private syncDisplayServersToUi() {
+    this.displayServerRepository.listServers().then((displayServers) => {
+      this.appRoot.serverList = displayServers;
+      if (!!this.displayServerBeingCreated) {
+        // Trigger Polymer array mutation.
+        this.appRoot.push('serverList', this.displayServerBeingCreated);
+      }
     });
   }
 
+  // Removes `displayServer` from the UI.
+  private removeServerFromDisplay(displayServer: DisplayServer) {
+    delete this.serverDisplayTable[displayServer.id];
+    this.displayServerRepository.removeServer(displayServer);
+    this.syncDisplayServersToUi();
+  }
+
+  // Shows the last server displayed, if there is one in local storage.
+  private maybeShowLastDisplayedServer() {
+    const lastDisplayedServerId = this.displayServerRepository.getLastDisplayedServerId();
+    if (!lastDisplayedServerId) {
+      return;  // No server was displayed when user quit the app.
+    }
+    const lastDisplayedServer = this.displayServerRepository.findServer(lastDisplayedServerId);
+    if (!lastDisplayedServerId) {
```

Comment:
Did you mean to check `lastDisplayedServer` here instead?

Reply:
Yes, good catch! Done in 8b3ab47